app/services/experiment_control_service.rb

```diff
@@ -55,6 +55,16 @@ def schedule_daily_stale_patient_notifications(name, patients_per_day: PATIENTS_
       experiment.running_state!
     end
 
+    def terminate_experiment(name)
+      experiment = Experimentation::Experiment.find_by!(name: name)
+      experiment.complete_state!
+
+      notifications = experiment.notifications.where(status: ["pending", "scheduled"])
+      notifications.find_each do |notification|
+        notification.status_cancelled!
+      end
+    end
```

Comment:
- Where do we expect this to be called?
- Will synchronously cancelling 1000s of notifications be slow, and do we need to think about adding a tx around this?
- Should we consider using the experiment's state to stop sending out notifications, and then marking them cancelled asynchronously?

Reply:
1) my expectation is that this would be called from the rails console if we ever need to end an experiment early. 
2) it might be a bit slow. I'm not sure what a transaction would do here. Are you suggesting we'd roll back all the changes if one of them fails? I think it would make more sense to raise a loud error, let someone fix the problem, then run it again from where they left off. 
3) i think a guard in the worker against the experiment being complete is a good idea. It's already guarded against the notification being cancelled and the feature flag being off, but there are edge cases where those two guards might not be sufficient. For example, if this job takes a long time and we have two experiments running simultaneously with the same feature flag. I'll add that. 
We could cancel the notifications async, but I'm not sure it's worth the additional complexity. As we saw today, background jobs are much harder to debug.

Reply:
This PR keeps getting put on the back burner. Seems like we're all good with points 1 and 2. So to address point 3... 

After Srihari's initial feedback, I decided to add the "cancelled" state option to the experiment model. That allows us to guard against sending notifications if the experiment is cancelled, which I believe handles Srihari's concerns that the job may take a while and could result in us sending notifications unintentionally because we hit the "abort" button but it couldn't cancel all the scheduled notifications before they could send. 

Maybe this helps to clarify. We have two basic kinds of experiments (not counting medication reminders). 
- current patients: 
  - start and end dates: don't really do anything. They're mostly used for historical reference.
  - `experiment.state_completed?` has no impact on the experiment. Can't have more than one "running" experiment of the same type at a time, so you do need to change an experiment to complete or canceled in order to run another active patient experiment.
  - `experiment.state_cancelled?` prevents scheduled notifications from being sent
- stale patients: 
  - start and end dates: used to guard against selecting more patients for the experiment. Selection only occurs if we're within the date range. End date is also used to mark the experiment as complete (see this PR, it's new functionality).
  - `experiment.state_completed?` daily selection does not occur if state is complete
  - `experiment.state_cancelled?` prevents scheduled notifications from being sent and also prevents daily selection from occurring. 
  
Does that clear anything up? It's all complicated by the fact that we have two different experiment types that run in very different ways.

Reply:
@kpethtel Understood, so cancelling an experiment needs to put an _immediate_ end to messages that were going to go out as part of the experiment, even if the `Notification`s have already been scheduled for the future.

I think this approach is 👌